File: Data Structures & Algorithms/binary-tree-diameter/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        int[] ans = new int[1];
        dfs(ans, root);
        return ans[0];
    }

    static int dfs(int[] ans, TreeNode root)
    {
        if (root == null)
            return 0; 

        int left = dfs(ans, root.left);
        int right = dfs(ans, root.right);

        ans[0] = Math.max(left + right, ans[0]);

        return 1 + Math.max(left, right); 
    }

  
}

File: Data Structures & Algorithms/string-encode-and-decode/submission-1.py
class Solution:

    def encode(self, strs: List[str]) -> str:
        chars = []
        for s in strs: 
            chars.append(str(len(s)))
            chars.append('#')
            chars.append(s)
        return "".join(chars) 

    def decode(self, s: str) -> List[str]:
        left, right = 0, 0
        ans = [] 
        while right < len(s): 
            while s[right] != '#':
                right +=1
            length = int(s[left:right])
            ans.append(s[right +1 :right + length + 1])
            left = right + length + 1
            right = right + length + 1

        return ans